### PR TITLE
Add RFC 1003: Mandatory human handle & device label

### PR DESCRIPTION
**Pull request by touilleMan**

### CODE REVIEW

File: docs/rfcs/1003-mandatory-human-handle.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+revoking `sgx-anti-virus-20230901` and replacing it by `sgx-anti-virus-20230902`).
+
+However in practice we never used non-human accounts.
+On top of that the human handle field doesn't prevent from creating non-human account:
+the email field can be set to a dummy value (or event to a valid address to contact
+the team responsible for this).
+
+### 3 - Hide identity from outsiders
+
+The Outsider profile is to be use for user that should be allowed to work in
```
Comment (FirelightFlagboy) [inline docs/rfcs/1003-mandatory-human-handle.md:35]:
```suggestion
The Outsider profile is to be used for user that should be allowed to work in
```

File: docs/rfcs/1003-mandatory-human-handle.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+an organization while not being trusted enough to know the full identity of
+who is part of said organization.
```
Comment (FirelightFlagboy) [inline docs/rfcs/1003-mandatory-human-handle.md:37]:
```suggestion
who is a part of said organization.
```

File: docs/rfcs/1003-mandatory-human-handle.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+1 - User 1 (user id = "alice") is created with an old Parsec version, hence it has no human handle
+2 - User 2 (user id = "bob") is created with human handle email = "alice@redacted.invalid"
+
+Now we end up with Parsec displaying the same human handle email for both user 1 and 2 :(
+The easy solution for this is simply to consider `redacted.invalid` as a reserved domain that
+is not allowed for real human handle.
+
+Typically `HumanHandle` method would look like:
+```
```
Comment (FirelightFlagboy) [inline docs/rfcs/1003-mandatory-human-handle.md:135]:
```suggestion
Typically `HumanHandle` method would look like:

```rust
```

File: docs/rfcs/1003-mandatory-human-handle.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    pub fn from_redacted_user(user_id: &UserID) -> Self {
+      // Here we compute label and email from the user id (hence email uses
+      // `redacted.invalid` domain)
+      // This method should only be used in the code deserializing `UserCertificate`
+    }
+```
+
+### 5 - Don't carry presence/absence of the fields to the GUI
+
+As we've seen, legacy certificate are handled just like redacted certificate. On top
```
Comment (FirelightFlagboy) [inline docs/rfcs/1003-mandatory-human-handle.md:151]:
```suggestion
As we've seen, legacy certificates are handled just like redacted certificates. On top
```

File: docs/rfcs/1003-mandatory-human-handle.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+of that redacted certificate are only provided to Outsider user.
```
Comment (FirelightFlagboy) [inline docs/rfcs/1003-mandatory-human-handle.md:152]:
```suggestion
of that redacted certificates are only provided to Outsider user.
```

File: docs/rfcs/1003-mandatory-human-handle.md [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+Given legacy certificate is not that common, there is no need to convey this information
```
Comment (FirelightFlagboy) [inline docs/rfcs/1003-mandatory-human-handle.md:154]:
```suggestion
Given legacy certificates are not that common, there is no need to convey this information
```

File: docs/rfcs/1003-mandatory-human-handle.md [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+to the GUI.
+
+Regarding redacted certificate, the GUI already has to know the user is an outsider
```
Comment (FirelightFlagboy) [inline docs/rfcs/1003-mandatory-human-handle.md:157]:
```suggestion
Regarding redacted certificates, the GUI already has to know the user is an outsider
```

File: docs/rfcs/1003-mandatory-human-handle.md [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+- Non-redacted certificate: `My PC` (i.e. device label field)
+- Redacted certificate: `e5904a3eccd8451b99ba09324104662e` (i.e. device name field, which is always a UUID)
+
+### 3 - Human handle for redacted certificate
+
+Just like for device label, human handle label is a superset of user id so the latter can
+be directly used.
+
+For the human handle email we have to be more cunning (especially given the email *is* the
+field used as identifier where label is only displayed to human with no other purpose
```
Comment (mmmarcos) [inline docs/rfcs/1003-mandatory-human-handle.md:110]: It is not clear to me why (or where) "the email is the field used as identifier", isn't the user id always use as identifier? 

Which makes me think, do we really need to bother at all with the email's local-part? How about always using something like `redacted@redacted.invalid`?

File: docs/rfcs/1003-mandatory-human-handle.md [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+of that redacted certificates are only provided to Outsider user.
+
+Given legacy certificates are not that common, there is no need to convey this information
+to the GUI.
+
+Regarding redacted certificates, the GUI already has to know the user is an outsider
+to adapt what is displayed (e.g. outsider cannot create new workspaces).
+
+Hence there is no need to provide a specific field to inform where the human handle /
+device label fields come from.
```
Comment (mmmarcos) [inline docs/rfcs/1003-mandatory-human-handle.md:202]: It seems to me that "redacted certificates are only provided to Outside user" is no longer true.
I understand what you mean (that this is currently the case), but with this proposed RFC, we will be 
actually switching to "redacted certificates are provided to the Outside users or Standard Legacy users".

Also, some rephrasing to avoid redundancy:
```suggestion
As we've seen, legacy certificates will be handled just like redacted certificates.

However, there is no need to convey this information (redacted or legacy) to the GUI.
Given legacy certificates are not that common, the GUI can adapt what is displayed
only based on user type: 

* Outside user (always redacted certificates): do not show human handle / device information
* Admin/Standard user (regular or redacted certificates): show human handle / device information
  *  If redacted certificate (because legacy user), it is acceptable to show the redacted information in this case

Hence there is no need to provide a specific field to inform where the human handle /
device label fields come from.
```

File: docs/rfcs/1003-mandatory-human-handle.md [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+
+What this also means is a user can register himself with an email in the `redacted.invalid`
+tld. This is considered ok given the email is validated by the greeter before the creation.
```
Comment (mmmarcos) [inline docs/rfcs/1003-mandatory-human-handle.md:205]: Isn't this phrase in contradiction with section 4 (`redacted.invalid not allowed for real human handle`)

File: docs/rfcs/1003-mandatory-human-handle.md (hunk 2 above)
Comment (FirelightFlagboy) [reply docs/rfcs/1003-mandatory-human-handle.md:37]: Event better <https://www.linguee.fr/francais-anglais/search?source=auto&query=who+is+part+of>

File: docs/rfcs/1003-mandatory-human-handle.md (hunk 8 above)
Comment (mmmarcos) [reply docs/rfcs/1003-mandatory-human-handle.md:111]: This is indeed interesting for the documentation (not sure if the user documentation though, or maybe dev/admin docs?)

Comment (touilleMan) [reply docs/rfcs/1003-mandatory-human-handle.md:111]: I would say this has it place in the regular: it explain a fundamental concept of Parsec that is useful for advanced end up (the kind of one that read documentation ^^)

typically we want to expose userid/deviceid/realmid in the GUI (in an advanced property section hidden by default) because it helps when working with server administrator

File: docs/rfcs/1003-mandatory-human-handle.md [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+(typically in case the account is compromised) while retaining his identity
+(e.g. user id "Gorge" is already taken by a now revoked user, hence you'll be
+known as "Jorge" from now on...).
+
+On the other hand a non-human account are fine with changing there id (typically
+revoking `sgx-anti-virus-20230901` and replacing it by `sgx-anti-virus-20230902`).
+
+However in practice we never used non-human accounts.
+On top of that the human handle field doesn't prevent from creating non-human account:
+the email field can be set to a dummy value (or event to a valid address to contact
```
Comment (FirelightFlagboy) [inline docs/rfcs/1003-mandatory-human-handle.md:30]:
```suggestion
the email field can be set to a dummy value (or even to a valid address to contact
```

File: docs/rfcs/1003-mandatory-human-handle.md [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+Typical use-case is a contractor working on a specific project (hence only having
+access to said project's workspace) while not actually part of the company the
+organization belongs to (hence the contractor cannot get the identity of all
+person working within the company by dumping the user certificates).
+
+This is achieved by providing the user and device certificates in two flavours:
+
+- regular one
+- redacted: which is the regular one, minus human handle / device label fields
+  which have been set to None
```
Comment (FirelightFlagboy) [inline docs/rfcs/1003-mandatory-human-handle.md:48]:
```suggestion
- redacted: i.e. Regular one without human handle / device label fields
```

It feels simple written this way

File: docs/rfcs/1003-mandatory-human-handle.md [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+
+It is annoying for the GUI to have to deal with the fact those fields might be missing,
+and in such case to roll a custom logic to still have something to display.
+
+This RFC propose to move the custom logic with the libparsec lower layer, so that
+the GUI has the guarantee those fields are always present from it point of view.
+
+## The solution
+
+tl;dr:
```
Comment (FirelightFlagboy) [inline docs/rfcs/1003-mandatory-human-handle.md:63]:
```suggestion
TL;DR:
```

File: docs/rfcs/1003-mandatory-human-handle.md [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+there is very few organization old enough to need legacy support, hence not having a
+specific way to handle not-redacted-but-also-no-human-handle-field is not worth it).
+
+For the username part we cannot use the user id field as-is, but instead apply to it
+the `uncaseify` algorithm that encode the case.
+
+`uncaseify` is as follow:
+
+1 - replace any `_` by `__`
+2 - replace any uppercase character by underscore followed by the lowercase version of the character
```
Comment (FirelightFlagboy) [inline docs/rfcs/1003-mandatory-human-handle.md:133]:
```suggestion
1. replace any `_` by `__`
2. replace any uppercase character by underscore followed by the lowercase version of the character
```

https://www.markdownguide.org/basic-syntax/#ordered-lists

File: docs/rfcs/1003-mandatory-human-handle.md [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+Finally considering the unicode, the good news is there is nothing to do: our email
+format already accept international email address (so a regular user can already use
+something like `关羽@蜀.汉`).
+
+### 4 - `redacted.invalid` not allowed for real human handle
+
+Let's consider the following scenario:
+
+1 - User 1 (user id = "alice") is created with an old Parsec version, hence it has no human handle
+2 - User 2 (user id = "bob") is created with human handle email = "alice@redacted.invalid"
```
Comment (FirelightFlagboy) [inline docs/rfcs/1003-mandatory-human-handle.md:171]:
```suggestion
1. User 1 (user id = "alice") is created with an old Parsec version, hence it has no human handle
2. User 2 (user id = "bob") is created with human handle email = "alice@redacted.invalid"
```

File: docs/rfcs/1003-mandatory-human-handle.md [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+Typically, `HumanHandle` method would look like:
+
+```rust
+impl HumanHandle {
+    pub fn new(email: &str, label: &str) -> Result<Self, &'static str> {
+      // Create regular human handle from the provided email & label
+      // here we ensure email doesn't use `redacted.invalid` domain
+    }
+
+    pub fn from_redacted_user(user_id: &UserID) -> Self {
```
Comment (FirelightFlagboy) [inline docs/rfcs/1003-mandatory-human-handle.md:186]:
```suggestion
    pub fn from_redacted_user(user_id: &UserID) -> Result<Self, &'static str> {
```

Will likely use `HumanHandle::new` since either way we will need to validate the email

File: docs/rfcs/1003-mandatory-human-handle.md [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+      // Here we compute label and email from the user id (hence email uses
+      // `redacted.invalid` domain)
+      // This method should only be used in the code deserializing `UserCertificate`
```
Comment (FirelightFlagboy) [inline docs/rfcs/1003-mandatory-human-handle.md:189]: If this method should only be used by `UserCertificate`, why not making it a method of `UserCertificate` ?

File: docs/rfcs/1003-mandatory-human-handle.md [hunk 18]
```diff
@@ -0,0 +1,235 @@
...
+### 6 - Optional device label / human handle turned mandatory in some schemas
+
+Multiple schemas contains an Optional human handle and/or device label.
+They can be divided into two categories:
+
+1. the option was present for future-proofing and never used in practice
+2. the option was present for backward compatibility
+
+Items in case 1 are:
+- command `invite_info`
```
Comment (FirelightFlagboy) [inline docs/rfcs/1003-mandatory-human-handle.md:216]:
```suggestion
Items in case 1 are:

- command `invite_info`
```

File: docs/rfcs/1003-mandatory-human-handle.md [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+- command `human_find` (removed in APIv4 so can be ignored)
+- data `invite_device_confirmation` & `invite_user_confirmation`
+- data `invite_device_data` & `invite_user_data`
+- data `pki_enrollment_answer_payload`
+- `device_file_recovery`
+- `device_file_password`
+- `device_file_smartcard`
+
+Items in case 2 are:
+- `legacy_device_file_password`
```
Comment (FirelightFlagboy) [inline docs/rfcs/1003-mandatory-human-handle.md:226]:
```suggestion
Items in case 2 are:

- `legacy_device_file_password`
```

File: docs/rfcs/1003-mandatory-human-handle.md (hunk 11 above)
Comment (touilleMan) [reply docs/rfcs/1003-mandatory-human-handle.md:30]: :+1:

File: docs/rfcs/1003-mandatory-human-handle.md (hunk 13 above)
Comment (touilleMan) [reply docs/rfcs/1003-mandatory-human-handle.md:63]: image

File: docs/rfcs/1003-mandatory-human-handle.md (hunk 14 above)
Comment (touilleMan) [reply docs/rfcs/1003-mandatory-human-handle.md:133]: :+1:

File: docs/rfcs/1003-mandatory-human-handle.md (hunk 15 above)
Comment (touilleMan) [reply docs/rfcs/1003-mandatory-human-handle.md:171]: :+1:

File: docs/rfcs/1003-mandatory-human-handle.md (hunk 16 above)
Comment (touilleMan) [reply docs/rfcs/1003-mandatory-human-handle.md:186]: We can always create a valid `HumanHandle` from a `UserID`, so no need to return an error

File: docs/rfcs/1003-mandatory-human-handle.md (hunk 17 above)
Comment (touilleMan) [reply docs/rfcs/1003-mandatory-human-handle.md:189]: This is also used when deserializing old `LocalDevice` and `LegacyDeviceFilePassword`

on top of that `HumanHandle::new` checks that `redacted.invalid` is not used, hence we need a dedicated method when we need to use it

File: docs/rfcs/1003-mandatory-human-handle.md (hunk 18 above)
Comment (touilleMan) [reply docs/rfcs/1003-mandatory-human-handle.md:216]: :+1:

File: docs/rfcs/1003-mandatory-human-handle.md (hunk 19 above)
Comment (touilleMan) [reply docs/rfcs/1003-mandatory-human-handle.md:226]: :+1:

**Review by FirelightFlagboy — APPROVED** [review]
